File: exercism/elixir/word-count/word_count.exs
defmodule Words do
  @doc """
  Count the number of words in the sentence.

  Words are compared case-insensitively.
  """

  @pattern ~r/\w+-?\w*/u

  @spec count(String.t) :: map
  def count(sentence) do
    sentence
    |> String.split
    |> do_count(%{})
  end

  defp do_count([], acc) do
    acc
  end

  defp do_count([word | tail], acc) do
    word
    |> apply_pattern
    |> sanitize
    |> reduce(acc)
    |> (&do_count(tail, &1)).()
  end

  defp apply_pattern(word) do
    case Regex.run(@pattern, word) do
      [valid] -> valid
      nil     -> nil
    end
  end

  defp sanitize(nil) do
    []
  end

  defp sanitize(word) do
    word
    |> String.downcase
    |> String.replace("_", " ")
    |> String.split
  end

  defp reduce([], acc) do
    acc
  end

  defp reduce([word | tail], acc) do
    tail |> reduce(Map.update(acc, word, 1, &inc/1))
  end

  defp inc(n), do: n + 1

end

File: exercism/clojure/word-count/src/word_count.clj
(ns word-count
  (:require [clojure.string :as str]))

(defn update-word
  [word acc]
  (let [valid (re-find #"\w+-?\w*" (str/lower-case word))]
  (if (nil? valid)
    acc
    (update acc valid (fnil inc 0)))))

(defn do-word-count
  [words acc]
  (let [word (first words)]
  (let [tail (rest words)]
  (if (empty? tail)
    (update-word word acc)
    (recur tail (update-word word acc))))))

(defn word-count
  [sentence]
  (let [words (str/split sentence #" ")]
  (do-word-count words {})))
